packages/sdk/src/reliability-score.ts

```diff
@@ -0,0 +1,157 @@
+import { utils } from "rpch-common";
+
+/**
+ * Possible `result` values.
+ * @type success: we have received an honest and valid response.
+ * @type dishonest: we have received a response but Kevlar says its not honest.
+ * @type failed: we have received no response for that request.
+ */
+export type Result = "success" | "dishonest" | "failed";
+
+export type ResponseMetric = {
+  id: string;
+  createdAt: Date;
+  result: Result;
+};
+
+export type Stats = {
+  success: number;
+  dishonest: number;
+  failed: number;
+};
+
+// ! Need to export these?
+export const FRESH_NODE_THRESHOLD = 20;
+export const MAX_RESPONSES = 100;
+
+/**
+ * Way to measure if the HOPRd entry node
+ * we are using is reliable or not.
+ */
+export default class ReliabilityScore {
+  /**
+   * Keeps track of metrics.
+   */
+  // ! metrics should be public or private?
+  public metrics = new Map<
+    string,
+    {
+      responses: Map<string, ResponseMetric>;
+      stats: Stats;
+      sent: number;
+      updatedAt: Date;
+    }
+  >();
+
+  /**
+   * Keeps track of calculated score.
+   * The `score` range goes from 0 to 1.
+   */
+  private score = new Map<string, number>();
+
+  /**
+   * Get a node's responses.result stats.
+   * @param peerId
+   * @returns object with number of successful, dishonest and failed responses.
+   */
+  private getResultsStats(peerId: string): Stats {
+    if (this.metrics.has(peerId)) {
+      const responses = Array.from(this.metrics.get(peerId)!.responses);
+      return responses.reduce(
+        (acc, [_, response]) => {
+          if (response.result === "success") {
+            acc.success++;
+          } else if (response.result === "dishonest") {
+            acc.dishonest++;
+          } else if (response.result === "failed") {
+            acc.failed++;
+          }
+          return acc;
+        },
+        { success: 0, dishonest: 0, failed: 0 }
+      );
+    } else return { success: 0, dishonest: 0, failed: 0 };
+  }
+
+  /**
+   * Add new metric to the metrics Map.
+   * @param peerId
+   * @param requestId
+   * @param result
+   */
+  public addMetric(peerId: string, requestId: string, result: Result) {
+    let nodeMetrics = this.metrics.get(peerId);
+
+    if (!nodeMetrics) {
+      nodeMetrics = {
+        responses: new Map<string, ResponseMetric>(),
+        stats: { success: 0, dishonest: 0, failed: 0 },
+        sent: 0,
+        updatedAt: new Date(),
+      };
+      this.metrics.set(peerId, nodeMetrics);
+    }
+
+    nodeMetrics.responses.set(requestId, {
+      id: `id${utils.generateRandomNumber()}`,
```

Comment:
this ID here should match the request id, but we don't need to include it at all